# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
data.groupby('children').agg({'debt':['count','sum','mean']})

debt                
          count   sum      mean
children                       
0         14091  1063  0.075438
1          4808   444  0.092346
2          2052   194  0.094542
3           330    27  0.081818
4            41     4  0.097561
5             9     0  0.000000

**Вывод:** 

Можно сказать, что наличие детей уменьшает вероятность возврата кредита в срок. Зависимость между количеством детей и возвратом кредита слабая. Выявлении зависимоти осложнено малым число займов с числом детей больше двух. 

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
#print(data.columns)
data.groupby('family_status').agg({'debt':['count','sum','mean']})

debt               
                       count  sum      mean
family_status                              
Не женат / не замужем   2796  273  0.097639
в разводе               1189   84  0.070648
вдовец / вдова           951   63  0.066246
гражданский брак        4134  385  0.093130
женат / замужем        12261  927  0.075606

**Вывод:** 

В таблице выше первый столбец отражает семейный статус, а последний долю просроченных платежей от числа займов. Зависимость между семейным положением и возвратом кредита в срок выражена хорошо. Можно отметить два категории лиц. Первая это те кто в не женаты либо имеют гражданский брак. Вторая это лица в разводе, вдовы/вдовцы и лица в браке. Вторая группа возращает в срок с большей вероятностью чем первая. 

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
data['total_income_qcut4']=pd.qcut(data['total_income'], 4)
data.groupby('total_income_qcut4').agg({'debt':['count','sum','mean']})

debt               
                      count  sum      mean
total_income_qcut4                        
(20666.999, 107507.0]  5333  427  0.080068
(107507.0, 142594.0]   5450  480  0.088073
(142594.0, 195842.0]   5216  444  0.085123
(195842.0, 2265604.0]  5332  381  0.071455

In [25]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
data.groupby('total_income_category').agg({'debt':['count','sum','mean']})

debt                
                       count   sum      mean
total_income_category                       
A                         25     2  0.080000
B                       5014   354  0.070602
C                      15921  1353  0.084982
D                        349    21  0.060172
E                         22     2  0.090909

**Вывод:** 

В таблице выше первый столбец отражает категорию дохода, а последний долю просроченных платежей от числа займов в данной категории.  Можно было сказать, что наименее и наиболее обеспеченные категории граждан более склонны  просрочкам ппо займам, но число таких данных не позволяет сделать достоверное заключение.  

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [26]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
data.groupby('purpose_category').agg({'debt':['count','sum','mean']})

debt               
                          count  sum      mean
purpose_category                              
операции с автомобилем     4279  400  0.093480
операции с недвижимостью  10751  780  0.072551
получение образования      3988  369  0.092528
проведение свадьбы         2313  183  0.079118

**Вывод:** 

В таблице выше первый столбец отражает цель займа, а последний долю просроченных платежей от числа займов с данной целью. 
Можно отметить что займы для операций с недвижимостью и проведения свадьбы позращаются чаще чем займы для получения образования или покупки автомобиля. 

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

In [27]:
data_org = pd.read_csv('/datasets/data.csv')
#data_org.info() 
data_org.loc[data_org['days_employed'].isna()].head(100)

children  days_employed  dob_years education  education_id  \
12          0            NaN         65   среднее             1   
26          0            NaN         41   среднее             1   
29          0            NaN         63   среднее             1   
41          0            NaN         50   среднее             1   
55          0            NaN         54   среднее             1   
..        ...            ...        ...       ...           ...   
959         0            NaN         25   среднее             1   
973         0            NaN         49   среднее             1   
974         0            NaN         60   среднее             1   
978         3            NaN         28   среднее             1   
980         0            NaN         53    высшее             0   

             family_status  family_status_id gender  income_type  debt  \
12        гражданский брак                 1      M    пенсионер     0   
26         женат / замужем                 0      M  госслужащий     0   
29   Не женат / не замужем                 4      F    пенсионер     0   
41         женат / замужем                 0      F  госслужащий     0   
55        гражданский брак                 1      F    пенсионер     1   
..                     ...               ...    ...          ...   ...   
959       гражданский брак                 1      M    сотрудник     0   
973        женат / замужем                 0      F    сотрудник     0   
974        женат / замужем                 0      M    сотрудник     0   
978       гражданский брак                 1      F    сотрудник     0   
980        женат / замужем                 0      F    сотрудник     0   

     total_income                           purpose  
12            NaN                   сыграть свадьбу  
26            NaN                       образование  
29            NaN  строительство жилой недвижимости  
41            NaN  сделка с подержанным автомобилем  
55            NaN                   сыграть свадьбу  
..            ...                               ...  
959           NaN                       образование  
973           NaN                 операции с жильем  
974           NaN                высшее образование  
978           NaN                      ремонт жилью  
980           NaN             получение образования  

[100 rows x 12 columns]

*Ответ:* 

Пропуски в столбце days_employed и total_income могли появится как случайность либо когда заемщик не знал что указать например если он пенсионер или компаньен, либо не хотел указывать нулевые значения. 

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 

Заполнение медианными значениями не влияет на медианное значение столбца, но при этом сохраняет другие данные которые могут оказать важными. 


<div style="background-color:black; color:#ffffff; text-align:center; padding:2%">
    
**Александр:** Медиана отличается от среднего математическим определением и совпадает только в случае симметричного распределения. В контексте замены для количественных переменных заполнение пропущенных значений медианным значением является предпочтичельным по следующим причинам 

1. медиана устойчива к выбросам. среднее значение может быть сильно искажено выбросами, 
 тогда как медиана менее чувствительна к выбросам и таким более репрезентативна для центральной части данных.

2. Заполнение пропусков медианным значением не искажает распределение данных. 
  Если бы мы заполнили пропуски средним значением, это могло бы привести к смещению распределения и повлиять на результаты анализа.

3. Заполнение медианными значениеми вычислительно более эффективно чем замена на среднии.


### Шаг 4: общий вывод.

**Вывод** 

Проведен анализ базы данны по заемщикам на предмет зависимости целевого значения (задержки/возрата по займу) от:
* количества дней трудоустройства, 
* кол-ва детей в семье
* семейного статуса 
* дохода заемщика 
* цели займа

Выявлены следущее недостатки в исходных данных

* пропуски в доходе и кол-ва дней трудоустройства
* одна и таже цель займа встречается в разных вариантах
* образование встречается в разных регистрах
* число дней трудоустройства -- встречаются отрицательные значения
* отрицательные значения и черезмерно высокие кол-во детей

    
При анализе исходной таблице были выялены пропуски в столбце total_income пропуски 
Полученые значения доли задержек покредиту в зависимости от кол-ва детей представлены на следующей таблице 

|  кол-во детей |   кол-во задержек в % |
|:-----------:|:------|
|          0 |   7.5% |
|          1 |   9.2% |
|          2 |   9.4% |
|          3 |   8.2% |
|          4 |   9.8% |
|          5 |    0%  |

Можно сказать, что наличие детей уменьшает вероятность возврата кредита в срок. Зависимость между количеством детей и возвратом кредита слабая. Выявлении зависимоти осложнено малым число займов с числом детей больше двух.

Полученые значения доли задержек покредиту в семейного статуса представлены на следующей таблице 

| семейный статус   |   кол-во задержек в %  |
|:----------------------|--------------:|
| Не женат / не замужем |          9.8% |
| в разводе             |          7.1% |
| вдовец / вдова        |          6.6% |
| гражданский брак      |          9.3% |
| женат / замужем       |          7.6% |

Зависимость между семейным положением и возвратом кредита в срок выражена хорошо. Можно отметить два категории лиц. Первая это те кто в не женаты либо имеют гражданский брак. Вторая это лица в разводе, вдовы/вдовцы и лица в браке. Вторая группа возращает в срок с большей вероятностью чем первая.


Проезведен расчет долей задержек по кредиту в зависимости от дохода заемщика. Результаты преведены на следующей таблице 

| интервал дохода   |   кол-во задержек в % |
|:----------------------|-------------------:|
| (20666.999, 107507.0] |     8.0% |
| (107507.0, 142594.0]  |     8.8% |
| (142594.0, 195842.0]  |     8.5% |
| (195842.0, 2265604.0] |     7.1% |

Из таблицы видно, что наименее и наиболее обеспеченные категории граждан более склонны просрочкам ппо займам.

| цель заема     | кол-во задержек в % |
|:-------------------------|-------------------:|
| операции с автомобилем   |          9.3% |
| получение образования    |          9.3% |
| операции с недвижимостью |          7.3% |
| проведение свадьбы       |          7.9% |


Разделяя заемщиком на категории по доходу, семейному положению, и кол-ву детей в семье можно получить оценку на вероятность возвата займа в срок. 

Наибольшую долю возвата наблюдается для

* операций с недвижимостью либо для свадьбы
* при среднем уровне дохода
* лиц в разводе, вдовы/вдовцы или в браке
* лиц без детей

Наименьшая доля возвата наблюдается для

* цели займа для операций с автомобилем либо для получения образования 
* при низком и высоком заработке заемщика 
* лиц которые не женаты либо в гражданском браке
* лиц с детьми


Важно отметить, что статистическая значимость вариаций доли задержек не изучалась в проведенном исследовании. 